src/alphafold_impact/pipelines/data_collection_oa/nodes.py

```diff
@@ -0,0 +1,188 @@
+            "concepts": paper["concepts"],
+            "keywords": paper["keywords"],
+            "grants": paper["grants"],
+        }
+        for paper in response
+    ]
+
+
+def _citation_works_generator(
+    mailto: str, perpage: str, work_id: str, direction: str
+) -> Iterator[list]:
+    """Creates a generator that yields a list of works from the OpenAlex API based on a 
+    given work ID.
+
+    Args:
+        work_id (str): A single work ID to filter by 'cites'.
+        direction (str): The direction of the citation. Either 'cites' or 'cited_by'.
+
+    Yields:
+        Iterator[list]: A generator that yields a list of works from the OpenAlex API 
+        based on a given work ID.
+    """
+
+    cursor_url = (
+        f"https://api.openalex.org/works?filter={direction}:{work_id}"
+        f"&mailto={mailto}&per-page={perpage}&cursor={{}}"
+    )
+
+    cursor = "*"
+    session = requests.Session()
+    retries = Retry(total=5, backoff_factor=0.3)
+    session.mount("https://", HTTPAdapter(max_retries=retries))
+
+    # make a call to estimate total number of results
+    response = session.get(cursor_url.format(cursor), timeout=20)
+    data = response.json()
+    total_results = data["meta"]["count"]
+    num_calls = total_results // int(perpage) + 1
+    logger.info(
+        "Total results for %s: %s, in %s calls", work_id, total_results, num_calls
+    )
+
+    while cursor:
+        response = session.get(cursor_url.format(cursor), timeout=20)
+        data = response.json()
+        results = data.get("results")
+        cursor = data["meta"].get("next_cursor", False)
+        yield results
+
+
+def collect_citation_papers(
+    mailto: str, perpage: str, work_ids: Union[str, List[str]], direction: str
+) -> dict:
+    """Collects all papers cited by specific work IDs.
+
+    Args:
+        work_ids (List[str]): A list of work IDs to filter by either 'cites' or 'cited_by'.
+        direction (str): The direction of the citation. Either 'cites' or 'cited_by'.
+
+    Returns:
+        dict: A dictionary containing the work_id as key and the list of papers as value.
+    """
+
+    assert direction in ["cites", "cited_by"], (
+        f"Invalid direction: {direction}. " f"Must be either 'cites' or 'cited_by'."
+    )
+
+    if isinstance(work_ids, str):
+        work_ids = [work_ids]
+
+    all_papers = {}
+    for work_id in work_ids:
+        papers_for_id = []
+        for page, papers in enumerate(
+            _citation_works_generator(
+                mailto=mailto, perpage=perpage, work_id=work_id, direction=direction
+            )
+        ):
+            papers_for_id.extend(_parse_results(papers))
+            logger.info(
+                "Fetching page %s of %s. Total papers collected: %s",
+                page,
+                work_id,
+                len(papers_for_id),
+            )
+        all_papers[work_id] = papers_for_id
+
+    return all_papers
```

Comment:
When running the kedro pipeline, does the output of this function gets saved as a partitioned dataset? Where the keys of the dict are the name for each partition and the values are the data it saves?

Reply:
Precisely!